### ISSUE TITLE
Stale NFS Handle

### ISSUE THREAD

**Issue by nick-schultz** [ccache/ccache#68]
We are sharing the ccache dir on an NFS share.  During high load, we are seeing consistent "Stale NFS file handle" errors to the stats file.

Any suggestions as to what can be done to avoid this problem? 

````
[2016-02-10T14:16:20.237532 12551] Trying direct lookup

[2016-02-10T14:16:20.238577 12551] Looking for object file hash in /nfs/compile_cache/a/5/c2da3efd19dcc3ef79042ca6df7c02-26442.manifest
[2016-02-10T14:16:20.706020 12551] Got object file hash from manifest

[2016-02-10T14:16:20.711564 12551] Copying /nfs/compile_cache/1/c/87c3236c570c5e35767209c40fcdc7-7031461.o to /out_dir/BubbleInsertionTest.obj via /out_dir/BubbleInsertionTest.obj.fmyciw202.12551.Jq5p9b (uncompressed)

[2016-02-10T14:16:20.724023 12551] Created from cache: /nfs/compile_cache/1/c/87c3236c570c5e35767209c40fcdc7-7031461.o -> /out_dir/BubbleInsertionTest.obj

[2016-02-10T14:16:20.726563 12551] Copying /nfs/compile_cache/1/c/87c3236c570c5e35767209c40fcdc7-7031461.d to /out_dir/BubbleInsertionTest.d via /out_dir/BubbleInsertionTest.d.fmyciw202.12551.kcTxOq (uncompressed)

[2016-02-10T14:16:20.734324 12551] Created from cache: /nfs/compile_cache/1/c/87c3236c570c5e35767209c40fcdc7-7031461.d -> /out_dir/BubbleInsertionTest.d

[2016-02-10T14:16:20.737888 12551] Succeeded getting cached result
[2016-02-10T14:16:20.738724 12551] Acquired lock /nfs/compile_cache/1/stats.lock
[2016-02-10T14:16:20.750046 12551] Releasing lock /nfs/compile_cache/1/stats.lock
[2016-02-10T14:16:20.750053 12551] Unlink /nfs/compile_cache/1/stats.lock
[2016-02-10T14:16:20.751956 12551] Result: cache hit (direct)             
[2016-02-10T14:16:20.751964 12551] Cleaning up cache directory /nfs/compile_cache/1
******[2016-02-10T14:16:56.050417 12551] FATAL: lstat /nfs/compile_cache/1/stats failed: Stale NFS file handle*****
```

**Comment by itensionanders** [comment]
Avoid NFS ? :-)

Seriously, it probably needs to take a read lock when accessing stats...
(or avoid trying to stat the "stats" files in traverse altogether, perhaps ?)

But you could have a look at dev/memcached, as an alternative to NFS.

**Comment by nick-schultz** [comment]
I also came across the traverse function as well. As mentioned in my pull request, I came across an early commit, 3049c10, that involved ignoring "No such file" cases.  Ignoring this error here may be applicable as well.

**Comment by itensionanders** [comment]
Sounds reasonable.

We had to turn off cleaning on our shared cache, due to the "thundering herd" problem when the cache grew full and *everyone* started cleaning/stat'ing the files at the same time... Cleaned it from cron instead.

**Comment by itensionanders** [comment]
Should be fixed with 5e31247ebe30b4a74053e0dfda834689e42abf91

**Comment by itensionanders** [comment]
Should be fixed in 3.2.5
